### PR TITLE
util: add table codec

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @coocood @shenli

File: src/util/codec/table.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    Ok(k)
+}
+
+/// `decode_handle` decodes the key and gets the handle.
+pub fn decode_handle(encoded: &[u8]) -> Result<i64> {
+    if !encoded.starts_with(TABLE_PREFIX) {
+        return Err(Error::InvalidDataType(format!("record key expected, but got {:?}", encoded)));
+    }
+
+    let mut left = &encoded[TABLE_PREFIX.len()..];
```
Comment (coocood) [inline src/util/codec/table.rs]: left -> remained

**Comment by ngaut** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

File: src/util/codec/table.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+use super::{number, Result, Error, Datum, datum};
+
+const ID_LEN: usize = 8;
+const PREFIX_LEN: usize = 1 + ID_LEN /*table_id*/ + 2;
+const RECORD_ROW_KEY_LEN: usize = PREFIX_LEN + ID_LEN;
+const TABLE_PREFIX: &'static [u8] = b"t";
+const RECORD_PREFIX_SEP: &'static [u8] = b"_r";
+const INDEX_PREFIX_SEP: &'static [u8] = b"_i";
+
+fn append_table_record_prefix(mut buf: &mut [u8], table_id: i64) -> Result<()> {
```
Comment (qiuyesuifeng) [inline src/util/codec/table.rs]: How about using build or create instead of append?

File: src/util/codec/table.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+fn append_table_index_prefix(mut buf: &mut [u8], table_id: i64) -> Result<()> {
+    try!(buf.write_all(TABLE_PREFIX));
+    try!(number::encode_i64(buf, table_id));
+    try!((&mut buf[8..]).write_all(INDEX_PREFIX_SEP));
+    Ok(())
+}
+
+/// `encode_row_key` encodes the table id and record handle into a byte array.
+pub fn encode_row_key(table_id: i64, encoded_handle: &[u8]) -> Result<Vec<u8>> {
+    let mut k = vec![0; RECORD_ROW_KEY_LEN];
```
Comment (qiuyesuifeng) [inline src/util/codec/table.rs]: `k` seems like a number variable like `i` or `j`, so why not using `key` instead?

File: src/util/codec/table.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+/// `decode_handle` decodes the key and gets the handle.
+pub fn decode_handle(encoded: &[u8]) -> Result<i64> {
+    if !encoded.starts_with(TABLE_PREFIX) {
+        return Err(Error::InvalidDataType(format!("record key expected, but got {:?}", encoded)));
+    }
+
+    let mut remaining = &encoded[TABLE_PREFIX.len()..];
+    try!(number::decode_i64(remaining));
+
+    if !remaining[8..].starts_with(RECORD_PREFIX_SEP) {
```
Comment (qiuyesuifeng) [inline src/util/codec/table.rs]: How about using ID_LEN?

Comment (BusyJay) [inline src/util/codec/table.rs]: This actually will be refactored in the next pr.

File: src/util/codec/table.rs (hunk 2 above)
Comment (BusyJay) [inline src/util/codec/table.rs]: No, it doesn't need to.

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by BusyJay** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
And also line 29 and 36？

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**